In [2]:
import pandas as pd
import pickle
import numpy as np

In [3]:
df = pickle.load(open('dataset_level2.pkl','rb'))

In [4]:
df

#BATTING TEAM 
#BOWLING TEAM
#CITY
#CURRENT_SCORE
#BALLS LEFT
#WICKETS LEFT
#CURRET RUN RATE
#LAST FIVE

match_id batting_team bowling_team  ball  runs player_dismissed     city  \
0           2    Australia    Sri Lanka   0.1     0                0      NaN   
1           2    Australia    Sri Lanka   0.2     0                0      NaN   
2           2    Australia    Sri Lanka   0.3     1                0      NaN   
3           2    Australia    Sri Lanka   0.4     2                0      NaN   
4           2    Australia    Sri Lanka   0.5     0                0      NaN   
..        ...          ...          ...   ...   ...              ...      ...   
121       964    Sri Lanka    Australia  19.3     1                0  Colombo   
122       964    Sri Lanka    Australia  19.4     0                0  Colombo   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva  Colombo   
124       964    Sri Lanka    Australia  19.6     2                0  Colombo   
125       964    Sri Lanka    Australia  19.7     1                0  Colombo   

                        venue  
0    Melbourne Cricket Ground  
1    Melbourne Cricket Ground  
2    Melbourne Cricket Ground  
3    Melbourne Cricket Ground  
4    Melbourne Cricket Ground  
..                        ...  
121       R Premadasa Stadium  
122       R Premadasa Stadium  
123       R Premadasa Stadium  
124       R Premadasa Stadium  
125       R Premadasa Stadium  

[63888 rows x 8 columns]

In [5]:
df.isnull().sum()

match_id               0
batting_team           0
bowling_team           0
ball                   0
runs                   0
player_dismissed       0
city                8548
venue                  0
dtype: int64

HERE WE CAN ABLE TO KNOW THAT IN CITY COLUMN 8548 CITY DATA IS NULL (MISSING)

In [6]:
df[df['city'].isnull()]['venue'].value_counts()

Dubai International Cricket Stadium        2969
Pallekele International Cricket Stadium    2066
Melbourne Cricket Ground                   1453
Sydney Cricket Ground                       749
Adelaide Oval                               498
Harare Sports Club                          372
Sharjah Cricket Stadium                     249
Sylhet International Cricket Stadium        128
Carrara Oval                                 64
Name: venue, dtype: int64

HERE WE CAN OBSERVE THAT THE FIRST WORD OF VENUE REPRESENTS THE CITY THEREFORE WE CAN EXTARCT IT TO ANOTHER COLUMN 

In [7]:
cities = np.where(df['city'].isnull(),df['venue'].str.split().apply(lambda x:x[0]),df['city'])  #HERE ARE COLLECTING THE FIRST WORD OF VENUE WHERE CITY IS NULL
#str.split results in seperating the words from the venue
#str.split().apply(lambda x:x[0]) results in collecting the first word from seperated words 

In [8]:
df['city'] = cities

In [9]:
df.isnull().sum()

match_id            0
batting_team        0
bowling_team        0
ball                0
runs                0
player_dismissed    0
city                0
venue               0
dtype: int64

NOW WE CAN SEE THERE ARE NO CITIES WHERE THERE IS NULL VALUE OR MISSING DATA 

In [10]:
df.drop(columns=['venue'],inplace=True)

In [11]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                0   
1           2    Australia    Sri Lanka   0.2     0                0   
2           2    Australia    Sri Lanka   0.3     1                0   
3           2    Australia    Sri Lanka   0.4     2                0   
4           2    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       964    Sri Lanka    Australia  19.3     1                0   
122       964    Sri Lanka    Australia  19.4     0                0   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva   
124       964    Sri Lanka    Australia  19.6     2                0   
125       964    Sri Lanka    Australia  19.7     1                0   

          city  
0    Melbourne  
1    Melbourne  
2    Melbourne  
3    Melbourne  
4    Melbourne  
..         ...  
121    Colombo  
122    Colombo  
123    Colombo  
124    Colombo  
125    Colombo  

[63888 rows x 7 columns]

NOW WE ARE ONLY CONSIDERING THOSE MATCHES WHICH OCCURS IN A CITY MANY TIMES, SAY AROUND ONLY CONSIDERING THOSE CITIES WHERE MIN 5 MATCHES ARE PLAYED

In [12]:
df['city'].value_counts() [df['city'].value_counts()> 600]

Colombo            4086
Mirpur             3420
Johannesburg       3331
Dubai              2969
Auckland           2532
Cape Town          2369
London             2106
Pallekele          2066
Barbados           2050
Sydney             1579
Melbourne          1574
Durban             1520
St Lucia           1487
Wellington         1350
Lauderhill         1238
Hamilton           1219
Centurion          1189
Manchester         1144
Abu Dhabi          1099
Mumbai              995
Nottingham          991
Southampton         987
Mount Maunganui     938
Chittagong          868
Kolkata             844
Lahore              748
Delhi               739
Nagpur              735
Chandigarh          627
Adelaide            623
Bangalore           620
St Kitts            618
Cardiff             615
Christchurch        614
Trinidad            611
Name: city, dtype: int64

WE HAVE EXTRACTED THE LIST OF CITIES WHERE ATLEAST 5 MATCHES ARE PLAYED 

In [13]:
eligible_cities = df['city'].value_counts() [df['city'].value_counts()> 600].index.tolist()

In [14]:
eligible_cities

['Colombo',
 'Mirpur',
 'Johannesburg',
 'Dubai',
 'Auckland',
 'Cape Town',
 'London',
 'Pallekele',
 'Barbados',
 'Sydney',
 'Melbourne',
 'Durban',
 'St Lucia',
 'Wellington',
 'Lauderhill',
 'Hamilton',
 'Centurion',
 'Manchester',
 'Abu Dhabi',
 'Mumbai',
 'Nottingham',
 'Southampton',
 'Mount Maunganui',
 'Chittagong',
 'Kolkata',
 'Lahore',
 'Delhi',
 'Nagpur',
 'Chandigarh',
 'Adelaide',
 'Bangalore',
 'St Kitts',
 'Cardiff',
 'Christchurch',
 'Trinidad']

In [12]:
df = df[df['city'].isin(eligible_cities)]

In [16]:
df['current_score'] = df.groupby('match_id').cumsum()['runs']

C:\Users\suyas\AppData\Local\Temp\ipykernel_13520\1598928411.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.cumsum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df['current_score'] = df.groupby('match_id').cumsum()['runs']


In [14]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                0   
1           2    Australia    Sri Lanka   0.2     0                0   
2           2    Australia    Sri Lanka   0.3     1                0   
3           2    Australia    Sri Lanka   0.4     2                0   
4           2    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       964    Sri Lanka    Australia  19.3     1                0   
122       964    Sri Lanka    Australia  19.4     0                0   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva   
124       964    Sri Lanka    Australia  19.6     2                0   
125       964    Sri Lanka    Australia  19.7     1                0   

          city  current_score  
0    Melbourne              0  
1    Melbourne              0  
2    Melbourne              1  
3    Melbourne              3  
4    Melbourne              3  
..         ...            ...  
121    Colombo            125  
122    Colombo            125  
123    Colombo            125  
124    Colombo            127  
125    Colombo            128  

[50501 rows x 8 columns]

we sepearated over and balls from ball column

In [15]:
df['over'] = df['ball'].apply(lambda x:str(x).split(".")[0])
df['ball_no'] = df['ball'].apply(lambda x:str(x).split(".")[1])

C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\1757844946.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['over'] = df['ball'].apply(lambda x:str(x).split(".")[0])
C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\1757844946.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['ball_no'] = df['ball'].apply(lambda x:str(x).split(".")[1])


In [16]:
df['balls_bowled'] = (df['over'].astype('int')*6) + df['ball_no'].astype('int')

C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\572800509.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['balls_bowled'] = (df['over'].astype('int')*6) + df['ball_no'].astype('int')


In [19]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                0   
1           2    Australia    Sri Lanka   0.2     0                0   
2           2    Australia    Sri Lanka   0.3     1                0   
3           2    Australia    Sri Lanka   0.4     2                0   
4           2    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       964    Sri Lanka    Australia  19.3     1                0   
122       964    Sri Lanka    Australia  19.4     0                0   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva   
124       964    Sri Lanka    Australia  19.6     2                0   
125       964    Sri Lanka    Australia  19.7     1                0   

          city  current_score over ball_no  balls_bowled  
0    Melbourne              0    0       1             1  
1    Melbourne              0    0       2             2  
2    Melbourne              1    0       3             3  
3    Melbourne              3    0       4             4  
4    Melbourne              3    0       5             5  
..         ...            ...  ...     ...           ...  
121    Colombo            125   19       3           117  
122    Colombo            125   19       4           118  
123    Colombo            125   19       5           119  
124    Colombo            127   19       6           120  
125    Colombo            128   19       7           121  

[50501 rows x 11 columns]

In [17]:
df['balls_left'] = 120 - df['balls_bowled']
df['balls_left'] = df['balls_left'].apply(lambda x:0 if x<0 else x)     #if there are some extras then value can be negative so we put condition that there some extras then consider 0 otherwise balls left

C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\3308979863.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['balls_left'] = 120 - df['balls_bowled']
C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\3308979863.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['balls_left'] = df['balls_left'].apply(lambda x:0 if x<0 else x)     #if there are some extras then value can be negative so we put condition that there some extras then consider 0 otherwise balls left


In [21]:
df

match_id batting_team bowling_team  ball  runs player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                0   
1           2    Australia    Sri Lanka   0.2     0                0   
2           2    Australia    Sri Lanka   0.3     1                0   
3           2    Australia    Sri Lanka   0.4     2                0   
4           2    Australia    Sri Lanka   0.5     0                0   
..        ...          ...          ...   ...   ...              ...   
121       964    Sri Lanka    Australia  19.3     1                0   
122       964    Sri Lanka    Australia  19.4     0                0   
123       964    Sri Lanka    Australia  19.5     0      DM de Silva   
124       964    Sri Lanka    Australia  19.6     2                0   
125       964    Sri Lanka    Australia  19.7     1                0   

          city  current_score over ball_no  balls_bowled  balls_left  
0    Melbourne              0    0       1             1         119  
1    Melbourne              0    0       2             2         118  
2    Melbourne              1    0       3             3         117  
3    Melbourne              3    0       4             4         116  
4    Melbourne              3    0       5             5         115  
..         ...            ...  ...     ...           ...         ...  
121    Colombo            125   19       3           117           3  
122    Colombo            125   19       4           118           2  
123    Colombo            125   19       5           119           1  
124    Colombo            127   19       6           120           0  
125    Colombo            128   19       7           121           0  

[50501 rows x 12 columns]

In [18]:
df['player_dismissed'] = df['player_dismissed'].apply(lambda x:0 if x=='0' else 1)
df['player_dismissed'] = df['player_dismissed'].astype('int')
df['player_dismissed'] = df.groupby('match_id').cumsum()['player_dismissed']
df['wickets_left'] = 10 - df['player_dismissed']


C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\4265656194.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['player_dismissed'] = df['player_dismissed'].apply(lambda x:0 if x=='0' else 1)
C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\4265656194.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['player_dismissed'] = df['player_dismissed'].astype('int')
C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\4265656194.py:3: FutureWarning: The default value of numeric_only in DataFrameGroupBy.cumsum i

In [19]:
df

match_id batting_team bowling_team  ball  runs  player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                 0   
1           2    Australia    Sri Lanka   0.2     0                 0   
2           2    Australia    Sri Lanka   0.3     1                 0   
3           2    Australia    Sri Lanka   0.4     2                 0   
4           2    Australia    Sri Lanka   0.5     0                 0   
..        ...          ...          ...   ...   ...               ...   
121       964    Sri Lanka    Australia  19.3     1                 8   
122       964    Sri Lanka    Australia  19.4     0                 8   
123       964    Sri Lanka    Australia  19.5     0                 9   
124       964    Sri Lanka    Australia  19.6     2                 9   
125       964    Sri Lanka    Australia  19.7     1                 9   

          city  current_score over ball_no  balls_bowled  balls_left  \
0    Melbourne              0    0       1             1         119   
1    Melbourne              0    0       2             2         118   
2    Melbourne              1    0       3             3         117   
3    Melbourne              3    0       4             4         116   
4    Melbourne              3    0       5             5         115   
..         ...            ...  ...     ...           ...         ...   
121    Colombo            125   19       3           117           3   
122    Colombo            125   19       4           118           2   
123    Colombo            125   19       5           119           1   
124    Colombo            127   19       6           120           0   
125    Colombo            128   19       7           121           0   

     wickets_left  
0              10  
1              10  
2              10  
3              10  
4              10  
..            ...  
121             2  
122             2  
123             1  
124             1  
125             1  

[50501 rows x 13 columns]

now we calculate current run rate 

In [20]:
df['crr'] = (df['current_score']*6)/df['balls_bowled']

C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\914577034.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['crr'] = (df['current_score']*6)/df['balls_bowled']


In [21]:
df

match_id batting_team bowling_team  ball  runs  player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                 0   
1           2    Australia    Sri Lanka   0.2     0                 0   
2           2    Australia    Sri Lanka   0.3     1                 0   
3           2    Australia    Sri Lanka   0.4     2                 0   
4           2    Australia    Sri Lanka   0.5     0                 0   
..        ...          ...          ...   ...   ...               ...   
121       964    Sri Lanka    Australia  19.3     1                 8   
122       964    Sri Lanka    Australia  19.4     0                 8   
123       964    Sri Lanka    Australia  19.5     0                 9   
124       964    Sri Lanka    Australia  19.6     2                 9   
125       964    Sri Lanka    Australia  19.7     1                 9   

          city  current_score over ball_no  balls_bowled  balls_left  \
0    Melbourne              0    0       1             1         119   
1    Melbourne              0    0       2             2         118   
2    Melbourne              1    0       3             3         117   
3    Melbourne              3    0       4             4         116   
4    Melbourne              3    0       5             5         115   
..         ...            ...  ...     ...           ...         ...   
121    Colombo            125   19       3           117           3   
122    Colombo            125   19       4           118           2   
123    Colombo            125   19       5           119           1   
124    Colombo            127   19       6           120           0   
125    Colombo            128   19       7           121           0   

     wickets_left       crr  
0              10  0.000000  
1              10  0.000000  
2              10  2.000000  
3              10  4.500000  
4              10  3.600000  
..            ...       ...  
121             2  6.410256  
122             2  6.355932  
123             1  6.302521  
124             1  6.350000  
125             1  6.347107  

[50501 rows x 14 columns]

In [22]:
groups = df.groupby('match_id')
match_ids = df['match_id'].unique()
last_five = []
for id in match_ids:
    last_five.extend(groups.get_group(id).rolling(window=30).sum()['runs'].values.tolist())

C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\2253008329.py:5: FutureWarning: Dropping of nuisance columns in rolling operations is deprecated; in a future version this will raise TypeError. Select only valid columns before calling the operation. Dropped columns were Index(['batting_team', 'bowling_team', 'city'], dtype='object')
  last_five.extend(groups.get_group(id).rolling(window=30).sum()['runs'].values.tolist())


In [23]:
last_five

[nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 43.0,
 44.0,
 45.0,
 44.0,
 42.0,
 43.0,
 41.0,
 41.0,
 41.0,
 42.0,
 42.0,
 39.0,
 38.0,
 39.0,
 38.0,
 39.0,
 40.0,
 37.0,
 41.0,
 42.0,
 42.0,
 44.0,
 44.0,
 40.0,
 39.0,
 37.0,
 35.0,
 31.0,
 27.0,
 26.0,
 31.0,
 31.0,
 30.0,
 31.0,
 31.0,
 32.0,
 31.0,
 31.0,
 36.0,
 37.0,
 37.0,
 37.0,
 37.0,
 36.0,
 36.0,
 36.0,
 36.0,
 35.0,
 32.0,
 32.0,
 31.0,
 29.0,
 29.0,
 31.0,
 32.0,
 32.0,
 32.0,
 33.0,
 36.0,
 37.0,
 37.0,
 39.0,
 41.0,
 40.0,
 41.0,
 41.0,
 42.0,
 44.0,
 39.0,
 43.0,
 45.0,
 48.0,
 48.0,
 48.0,
 49.0,
 49.0,
 49.0,
 51.0,
 51.0,
 51.0,
 52.0,
 52.0,
 53.0,
 53.0,
 53.0,
 54.0,
 53.0,
 52.0,
 50.0,
 53.0,
 47.0,
 46.0,
 45.0,
 46.0,
 47.0,
 49.0,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan,
 nan

In [24]:
df['last_five'] = last_five

C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\1595651733.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['last_five'] = last_five


In [25]:
df

match_id batting_team bowling_team  ball  runs  player_dismissed  \
0           2    Australia    Sri Lanka   0.1     0                 0   
1           2    Australia    Sri Lanka   0.2     0                 0   
2           2    Australia    Sri Lanka   0.3     1                 0   
3           2    Australia    Sri Lanka   0.4     2                 0   
4           2    Australia    Sri Lanka   0.5     0                 0   
..        ...          ...          ...   ...   ...               ...   
121       964    Sri Lanka    Australia  19.3     1                 8   
122       964    Sri Lanka    Australia  19.4     0                 8   
123       964    Sri Lanka    Australia  19.5     0                 9   
124       964    Sri Lanka    Australia  19.6     2                 9   
125       964    Sri Lanka    Australia  19.7     1                 9   

          city  current_score over ball_no  balls_bowled  balls_left  \
0    Melbourne              0    0       1             1         119   
1    Melbourne              0    0       2             2         118   
2    Melbourne              1    0       3             3         117   
3    Melbourne              3    0       4             4         116   
4    Melbourne              3    0       5             5         115   
..         ...            ...  ...     ...           ...         ...   
121    Colombo            125   19       3           117           3   
122    Colombo            125   19       4           118           2   
123    Colombo            125   19       5           119           1   
124    Colombo            127   19       6           120           0   
125    Colombo            128   19       7           121           0   

     wickets_left       crr  last_five  
0              10  0.000000        NaN  
1              10  0.000000        NaN  
2              10  2.000000        NaN  
3              10  4.500000        NaN  
4              10  3.600000        NaN  
..            ...       ...        ...  
121             2  6.410256       32.0  
122             2  6.355932       32.0  
123             1  6.302521       32.0  
124             1  6.350000       33.0  
125             1  6.347107       32.0  

[50501 rows x 15 columns]

In [26]:
final_df = df.groupby('match_id').sum()['runs'].reset_index().merge(df,on='match_id')

C:\Users\suyas\AppData\Local\Temp\ipykernel_9416\3687880561.py:1: FutureWarning: The default value of numeric_only in DataFrameGroupBy.sum is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  final_df = df.groupby('match_id').sum()['runs'].reset_index().merge(df,on='match_id')


In [27]:
final_df = final_df[['batting_team','bowling_team','city','current_score','balls_left','wickets_left','crr','last_five','runs_x']]

In [28]:
final_df.dropna(inplace=True)

In [29]:
final_df.isnull().sum()

batting_team     0
bowling_team     0
city             0
current_score    0
balls_left       0
wickets_left     0
crr              0
last_five        0
runs_x           0
dtype: int64

In [56]:
final_df.dropna(inplace=True)

now there should be no bias so we are shuff;ing the data

In [30]:
final_df = final_df.sample(final_df.shape[0])

In [31]:
final_df.sample(2)

batting_team bowling_team          city  current_score  balls_left  \
45119      England     Pakistan         Dubai             56          66   
10641  New Zealand      England  Christchurch            138          11   

       wickets_left       crr  last_five  runs_x  
45119             7  6.222222       37.0     160  
10641             6  7.596330       45.0     153

TRAINING OF DATA

In [32]:
X = final_df.drop(columns=['runs_x'])
y = final_df['runs_x']
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.2,random_state=1)

In [33]:
X_train

batting_team  bowling_team             city  current_score  balls_left  \
45        Australia     Sri Lanka        Melbourne             55          75   
3798    New Zealand   West Indies  Mount Maunganui             80          72   
39429      Pakistan     Australia           Mirpur            180           6   
41978  South Africa   West Indies     Johannesburg             80          76   
36396     Australia     Sri Lanka           Sydney             35          78   
...             ...           ...              ...            ...         ...   
31103       England      Pakistan        Abu Dhabi            122           3   
35559      Pakistan  South Africa        Centurion             91          63   
49650      Pakistan         India           Mirpur             69          32   
12015       England     Australia      Southampton             66          63   
36736   West Indies    Bangladesh           Mirpur             54          72   

       wickets_left        crr  last_five  
45               10   7.333333       40.0  
3798             10  10.000000       46.0  
39429             6   9.473684       49.0  
41978             8  10.909091       57.0  
36396             8   5.000000       27.0  
...             ...        ...        ...  
31103             4   6.256410       36.0  
35559             9   9.578947       56.0  
49650             3   4.704545       23.0  
12015             8   6.947368       45.0  
36736             8   6.750000       29.0  

[30781 rows x 8 columns]

In [34]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestRegressor
from xgboost import XGBRegressor
from sklearn.metrics import r2_score,mean_absolute_error

In [35]:
trf = ColumnTransformer([
    ('trf',OneHotEncoder(sparse=False,drop='first'),['batting_team','bowling_team','city'])
]
,remainder='passthrough')

In [36]:
pipe = Pipeline(steps=[
    ('step1',trf),
    ('step2',StandardScaler()),
    ('step3',XGBRegressor(n_estimators=1000,learning_rate=0.2,max_depth=12,random_state=1))
])

In [37]:
pipe.fit(X_train,y_train)
y_pred = pipe.predict(X_test)
print(r2_score(y_test,y_pred))
print(mean_absolute_error(y_test,y_pred))

c:\Users\suyas\AppData\Local\Programs\Python\Python310\lib\site-packages\sklearn\preprocessing\_encoders.py:808: FutureWarning: `sparse` was renamed to `sparse_output` in version 1.2 and will be removed in 1.4. `sparse_output` is ignored unless you leave `sparse` to its default value.
  warnings.warn(


0.9879261270176096
1.6238220189068768


In [ ]:
pickle.dump(pipe,open('pipe.pkl','wb'))